contrib/internal/httptrace/httptrace.go

```diff
@@ -90,8 +90,31 @@ func ippref(s string) *netaddr.IPPrefix {
 	return nil
 }
 
+// genClientIPSpanTags generates the client IP related key/value map of tags that need to be added to the span.
+func genClientIPSpanTags(r *http.Request) map[string]string {
+	tags := map[string]string{}
+	ip, matches := getClientIP(r)
+	if matches == nil {
+		if ip.IsValid() {
+			tags[ext.HTTPClientIP] = ip.String()
+		}
+		return tags
+	}
+	sb := strings.Builder{}
+	for hdr, ip := range matches {
+		tags[ext.HTTPRequestHeaders+"."+hdr] = ip
+		sb.WriteString(hdr + ":" + ip + ",")
+	}
+	hdrList := sb.String()
+	tags[ext.MultipleIPHeaders] = hdrList[:len(hdrList)-1]
+	return tags
+}
+
 // getClientIP attempts to find the client IP address in the given request r.
-func getClientIP(r *http.Request) netaddr.IP {
+// If several IP headers are present in the request, the returned IP is invalid and the map gets filled with the
+// header/ip pairs for all IP headers found in the request. Otherwise, the returned map is nil.
+// See https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2118779066/Client+IP+addresses+resolution
+func getClientIP(r *http.Request) (netaddr.IP, map[string]string) {
```

Comment:
Let's refactor all of this.

The way it's written now, we build one map of header name -> value, then build another of tag name -> value, then finally converts that map to a list of start options for the span. 

Instead, let's just have one function that returns the tag set that can be `append`ed in `StartRequestSpan`.

Reply:
I think it's good to keep a function which retrieves the IP without worrying about start options, we may want to retrieve the IP for something other than span tags in the future. What about keeping `getClientIP` as is and changing `genClientIPSpanTags` to generate the start options directly. WDYT?

Reply:
I mean that splitting the part that retrieves the IP from the one that processes it into span tags makes sense to my eyes since retrieving the IP seems like something worth having on its own. I totally agree that doing it all in a single function is possible.

Reply:
We should write functions that solve the problems we have. For instance, if we want a list of `ddtrace.StartSpanOption`s that contain tags for the client IP's, it's generally better to have one function that does that than 2 that each do half the job.

Some reasons to break up a function are
1. if that function becomes too large to be comprehensible
2. if you want to reuse parts of the functionality
3. for testing purposes.

There are more, but these are big ones.

Right now, I think reducing the functions will simplify the code, so #1 does not apply, I think.

For no. 2, we may eventually want to use retrieve the IP for something else, but we may not, or we may want it in a different form. Until we actually need that functionality, it doesn't need to be separated. Cleaning this up and simplifying it will make it easier to pull out that logic into its own function later if we ever do need it elsewhere.

For no. 3, you could separate them, but the `getClientIP` functionality will be tested through `genClientIPSpanTags` anyway. If they are merged, that behavior is still easily covered by tests.

Whether or not we have 1 or 2 functions is not as important to me as that the code is simple and easy to understand. Currently the path from headers to tags is convoluted unnecessarily, as they make their way through several stages of maps and transformations, when we really just need one transformation from `*http.Request` to `[]StartSpanOption`.

Reply:
Refactored in 5f702117d6, let me know what you think